proposals/list-patterns.md

```diff
@@ -0,0 +1,142 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+positional_pattern
+  : type? positional_pattern_clause length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern
+  : type? length_pattern_clause property_or_list_pattern_clause? simple_designation?
+  ;
+
+list_pattern
+  : type? list_pattern_clause simple_designation?
+  ;
+
+property_pattern
+  : type? property_pattern_clause simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
```

Comment:
I'm not sure whether this should be `negated_pattern` or `pattern`. For example, should `..{ 1, 2 } or { 1, 2, 3 }` be allowed, or would that require parens `..({ 1, 2 } or { 1, 2, 3 })`?

Reply:
I'm open to both. I started this as an unary operator hence the current limitation. If you look at it as a kind of *prefix* we could possibly allow a general pattern instead. I do believe parens make that more clear but we probably shouldn't *force* it. Either way, something like `.. p or .. q` would be an error just for different reasons.

Reply:
Let's keep `negated_pattern` for now and add an open question. Thanks